### PR TITLE
TASK: improve performance of `getRelativeNodePathByUriPathSegmentProperties`

### PR DESCRIPTION
**Pull request by dimaip**
Use `NodeSearchService` to query for nodes with given uriPathSegment.

On pages with a lot of nodes on the same level (>1000) this drastically
improves performance.

### CODE REVIEW

**Comment by dimaip** [comment]
oups, forgot to adjust the tests, will do it soonish :-1:

File: TYPO3.Neos/Classes/TYPO3/Neos/Routing/FrontendNodeRoutePartHandler.php [hunk 1]
```diff
@@ -384,19 +398,16 @@ protected function resolveRoutePathForNode(NodeInterface $node)
...
-            foreach ($node->getChildNodes('TYPO3.Neos:Document') as $node) {
-                /** @var NodeInterface $node */
-                if ($node->getProperty('uriPathSegment') === $pathSegment) {
-                    $relativeNodePathSegments[] = $node->getName();
-                    $foundNodeInThisSegment = true;
-                    break;
-                }
-            }
-            if (!$foundNodeInThisSegment) {
+            $nodes = $this->nodeSearchService->findByProperties(['uriPathSegment' => $pathSegment], [$documentNodeType], $context, $node);
```
Comment (kdambekalns) [inline TYPO3.Neos/Classes/TYPO3/Neos/Routing/FrontendNodeRoutePartHandler.php]: This does work on the direct children only, the old code does…

Comment (dimaip) [inline TYPO3.Neos/Classes/TYPO3/Neos/Routing/FrontendNodeRoutePartHandler.php]: Damn. Any ideas how to fix it? Filter the results?

Comment (dimaip) [inline TYPO3.Neos/Classes/TYPO3/Neos/Routing/FrontendNodeRoutePartHandler.php]: Anyways, I've added a filter for that. Could be nice if there was a parameter to `findByProperties` to only consider direct children, WDYT?

**Comment by dimaip** [comment]
I give up on adjusting unit tests for this one.
My opinion: it doesn't make sense to unit test such huge and messy class as FrontentNodeRoutePartHandler, especially after this change, where part of essential logic is moved into the search service. I would propose to drop unit tests for it altogether until it's rewritten and just improve functional tests coverage for now (if needed).

Do you have better ideas?
Please help!
/cc @kdambekalns @kitsunet

**Comment by kitsunet** [comment]
First of all the syntax error should be fixed ;)
I am not a big fan of our tests but it's all we have, especially if we want to refactor, so 👎 for removing tests.

**Comment by dimaip** [comment]
@kitsunet oups, side-effects of night coding...
OK, I will try to spend some time to understand how to refactor unit tests to work with this change. The problem here is that I have replaced the crucial part of this class by for external service, and it would not be easy to mock it in the right way for me. Guess it's a good place to learn.

**Comment by kitsunet** [comment]
You are not at the sprint, right? Lets see, maybe I can have a look

**Comment by dimaip** [comment]
Unfortunately not this time :(
I'll try to do something tonight, and would be happy if you would continue where I left off.

**Comment by dimaip** [comment]
@kitsunet @kdambekalns okay guys, it wasn't as scary as I thought, I think I have fixed the tests. Please give it a look when you have time.

**Comment by hlubek** [comment]
@dimaip Same here, it would be nice to get this into 2.3 (if possible)

**Comment by kitsunet** [comment]
Would be nice, but I am not really confident to merge this into 2.3 yet.

File: TYPO3.Neos/Classes/TYPO3/Neos/Routing/FrontendNodeRoutePartHandler.php (hunk 1 above)
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Routing/FrontendNodeRoutePartHandler.php]: same issue as https://github.com/neos/neos-development-collection/pull/673#discussion_r76787221

**Comment by dimaip** [comment]
I'm testing performance now and noticed a really weird thing: if I take the current master, without applying my changes, it works really fast in the live workspace (0.18s for resolving a node with 7000 siblings), but it slows to a crawl when testing from the BE (3.48s). Can anyone make sense out of why it happens?

**Comment by dimaip** [comment]
Ah so I see that `getChildNodes` has its own layer of caching: `$this->context->getFirstLevelNodeCache()->getChildNodesByPathAndNodeTypeFilter`
So it's super fast once it's cached.

**Comment by dimaip** [comment]
Back to @aertmann's hint. Here are the performance metrics (all measured in backend):

| Metric type | master | this PR |
|----------------|-----------|-----------|
| 10 nesting levels | 0.025s | **1.13s** |
| 7000 siblings | **3.78s** | 0.22s |

**Comment by dimaip** [comment]
Anyways,  I've AOPed myself a fix tailored for my current project: https://github.com/psmb/PsmbDistr/commit/33375aa8a60db43574f30416f87a54c93b0878ce

I will pause project on this PR until somebody tells me which way to go. I'm stuck and have no idea how to fix it universally.

**Comment by jonnitto** [comment]
Ping @kdambekalns and @dimaip

**Pull request closed without merging**